### PR TITLE
Proposing Feature Dependencies

### PR DESCRIPTION
**Pull request by theluk**
The main purpose for this feature is to be able to invalidate cache entries and automatically invalidate other cache entries that are dependent from each other. Generally "Grouping" and "Namespacing" are good ideas but not suitable for the most use cases. 

### Problem

Let´t take the example of a simple one-to-one relation.
We have a **Brand** and a **Product**. Both are models in some relational database. On the website we want to build a product presentation that looks like this

    $view = sprintf("%s by %s", $product, $brand);

Now if we want to cache the view, currently neither namespacing nor grouping could help us invalidate the view, if it was cached and either product or brand changed. 

a general approach to cache all the items would be

    cache($product, "models/product/1")
    cache($brand, "models/brand/1")
    cache($view, "view/1");

if than a `$product` or `$brand` changes, both need to now exactly who used them, to be able to invalidate also `$view` as it would otherwise display invalid out-dated data.

### Solution

I will call my solution **Dependency Inheritance and Cache Invalidation**. 
Basically the `Stash\Interfaces\ItemInterface` gets one new method `addDependency`

    /**
     * Adds an ItemInterface as dependency. If the added
     * item gets invalidated, the caller also gets invalidated.
     * If $inherit is true, hierarchical dependencies are enabled.
     *
     * @param ItemInterface $dependency
     * @param boolean $inherit if dependencies should be inherited
     * @return boolean
     */
    public function addDependency(ItemInterface $dependency, $inherit = true);

A dependency is nothing more than an instance of `Stash\Interfaces\ItemInterface`. 
Therefore it is able to hierarchically depend on many different items. This enables us to invalidate one deeply nested cache item. This would also invalidate the root element, but not all untouched siblings. So that the root has to refetch only really missing cache items and the rest can be used from already cached items. 

Under the hood my solution uses the same technique as the grouping mechanism. It takes a cache key and combines it with the dependent in the following way

    $key = array_merge([$dependency->key, $this->key])
    // ==> model/foo/1/model/bar/2

This key is saved with a value of `1` the same way the `stampede protection` is saved. Then the new dependency `$key` is added to a list of dependencies into the items record array. 

The next time, when a cache item is fetched, the validation process also gets the dependency keys, fetches them using a `multiGet` and checks if the response count is the same as the count of the dependency keys. If any model that was saved and invalidated, everything in its group (`Model/Product/ID/*`) gets also deleted. This way not all dependency keys would be fetched and the count would mismatch. Bingo, we invalidated everything that depends on that item.

So the example from above would look like this

    Cache Key : Model/Product/1
    Attributes: {
                    "name": "My Product"
                }

    Cache Key : Model/Brand/1
    Attributes: {
                    "name": "Great Company"
                }

    Cache Key : View/ProductPresentation/1
    Attrbiutes: {
                    "text": render(sprintf("%s by %s", cache("Model/Product/1"), cache("Model/Brand/1")))
                }
    Dependency: [
                    "Model/Product/1/View/ProductPresentation/1",
                    "Model/Brand/1/View/ProductPresentation/1"
                ]

### CODE REVIEW

**Comment by tedivm** [comment]
I've done a quick read through and in general like the idea. I'll give it a further review once the test suite is passing.

**Comment by andrerom** [comment]
Interesting approach. Personally more a fan of tag based invalidation such as the one you can find in Symfony Cache Component _(cache based storage of cache tags)_, and unrelated code wise in Drupal 8 _(afaik db based persistance of cache tags)_, tab based invalidation can also be found other tools such as FOS HTTPCache / Varnish xkey, ...

It will let you tag your view cache with `[ 'product-1', 'brand-1']` and then typically in for instance decorator or event listener for updates to products you would invalidate cache for `"product-$id"`.

Benefit is:
- Only one call to memcached/redis is made ion lookups (with hierarchical cache in Stash it's _key parts +1_ , aka `my/product/1` = 4 cache backend lookups)
- More of a missing feature thing in Stash, but; Furthermore allows multi get lookups to result in still just one call to memcached/redis amplifying effect of point above. On setups with latency to cache, like on AWS, the benefit of all this can be quite huge

So basically the heavy lifting is done on invalidation, while lookup is fast.

**Comment by tedivm** [comment]
I've rebuild the test suite to work on Github- please rebase against the main branch and reopen this pull request if you're still interested in having it merged. Thank you!
